docs/contributor/reference/entity-lifecycle.md

```diff
@@ -28,6 +27,249 @@ There are two fundamental truths in this system:
+new relations identical to the destroyed ones, until those entities have
+finally been removed.
+
+In rough order of complexity, here's what happens when each entity kind is
+destroyed. Note that in general the appropriate action is contingent on
+mutable remote state, and many operations must be expressed as a transaction
+involving several documents: the state API must be prepared to handle aborted
+transactions and either diagnose definite failure or retry until the operation
+succeeds (or, perhaps, finally error out pleading excessive contention).
+
+
+### `juju remove-machine`
+
+
+Removing a machine involves a single transaction defined as follows:
+
+  * If the machine is not Alive, abort without error.
+  * If the machine is the last one with JobManageModel, or has any assigned
+    units, abort with an appropriate error.
+  * Set the machine to Dying.
+
+When a machine becomes Dying, the following operation occurs:
+
+  * The machine's agent sets the machine to Dead.
+
+When a machine becomes Dead, the following operations occur:
+
+  * The machine's agent terminates itself and refuses to run again.
+  * A Provisioner (a task running in some other machine agent) observes the
+    death, decommissions the machine's resources, and removes the machine.
+
+Removing a machine involves a single transaction defined as follows:
+
+  * If the machine is not Dead, abort with an appropriate error.
+  * Delete the machine document.
+
+
+### `juju remove-unit`
+
+Removing a unit involves a single transaction defined as follows:
+
+  * If the unit is not Alive, abort without error.
+  * Set the unit to Dying.
+
+When a unit becomes Dying, the following operations occur:
+
+  * The unit's agent leaves the scopes of all its relations. Note that this is
+    a potentially complex sequence of operations and may take some time; in
+    particular, any hooks that fail while the unit is leaving relations and
+    stopping the charm will suspend this sequence until resolved (just like
+    when the unit is Alive).
+  * The unit's agent then sets the unit to Dead.
+
+When a unit becomes Dead, the following operations occur:
+
+  * The unit's agent terminates itself and refuses to run again.
+  * The agent of the entity that deployed the unit (that is: a machine agent,
+    for a principal unit; or, for a subordinate unit, the agent of a principal
+    unit) observes the death, recalls the unit, and removes it.
+
+Removing a unit involves a single transaction, defined as follows:
+
+  * If the unit is a principal unit, unassign the unit from its machine.
+  * If the unit is a subordinate unit, unassign it from its principal unit.
+  * Delete the unit document.
+  * If its application is Alive, or has at least two units, or is in at least
+    one relation, decrement the application's unit count; otherwise remove the
+    application.
+
+
+### `juju remove-relation`
+
+Removing a relation involves a single transaction defined as follows:
```

Comment:
We should work together to revise these for Juju 4 as I implement them. Almost done with relations now.